crates/circuit/src/parameter/parameter_expression.rs

```diff
@@ -0,0 +1,2058 @@
+                return Err(ParameterError::UnknownParameters(different_symbols));
+            }
+        }
+
+        for (name, symbol) in self.name_map.iter() {
+            // check if the symbol will get replaced
+            if let Some(replacement) = map.get(symbol) {
+                // If yes, update the name_map. This also checks for duplicates.
+                for (replacement_name, replacement_symbol) in replacement.name_map.iter() {
+                    if let Some(duplicate) = name_map.get(replacement_name) {
+                        // If a symbol with the same name already exists, check whether it is
+                        // the same symbol (fine) or a different symbol with the same name (conflict)!
+                        if duplicate != replacement_symbol {
+                            return Err(ParameterError::NameConflict);
+                        }
+                    } else {
+                        // SAFETY: We know the key does not exist yet.
+                        unsafe {
+                            name_map.insert_unique_unchecked(
+                                replacement_name.clone(),
+                                replacement_symbol.clone(),
+                            )
+                        };
+                    }
+                }
+
+                // If we got until here, there were no duplicates, so we are safe to
+                // add this symbol to the internal replacement map.
+                symbol_map.insert(symbol.clone(), replacement.expr.clone());
+            } else {
+                // no replacement for this symbol, carry on
+                match name_map.entry(name.clone()) {
+                    Entry::Occupied(duplicate) => {
+                        if duplicate.get() != symbol {
+                            return Err(ParameterError::NameConflict);
+                        }
+                    }
+                    Entry::Vacant(e) => {
+                        e.insert(symbol.clone());
+                    }
+                }
+            }
+        }
+
+        let res = self.expr.subs(&symbol_map);
+        Ok(Self {
+            expr: res,
+            name_map,
+        })
+    }
+
+    /// Bind symbols to values.
+    ///
+    /// # Arguments
+    ///
+    /// * map - A hashmap with [Symbol] keys and [Value]s to replace these
+    ///     symbols with.
+    /// * allow_unknown_parameter - If `false`, returns an error if any symbol in the
+    ///     hashmap is not present in the expression. If `true`, unknown symbols are ignored.
+    ///     Setting to `true` is slightly faster as it does not involve additional checks.
+    ///
+    /// # Returns
+    ///
+    /// * `Ok(Self)` - A parameter expression with the bound symbols.
+    /// * `Err(ParameterError)` - An error if binding failed.
+    pub fn bind(
+        &self,
+        map: &HashMap<Symbol, Value>,
+        allow_unknown_parameters: bool,
+    ) -> Result<Self, ParameterError> {
```

Comment:
If you wanted to allow any sort of mapping between `Symbol` and `Value` you can change the signature slightly here:


```rs
    pub fn bind(
        &self,
        map: I,
        allow_unknown_parameters: bool,
    ) -> Result<Self, ParameterError> 
    where:
        I: IntoIter<Item = (Symbol, Value)>,
    {
```

Though it seems that here the mapping is borrowed and not owned, so this wouldn't work as well.